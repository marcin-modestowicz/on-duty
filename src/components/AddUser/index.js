//@flow
export { default } from "./AddUser";
