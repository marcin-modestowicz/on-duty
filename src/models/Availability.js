// @flow
import { observable, computed, action } from "mobx";

export const AVAILABILITY_STATUSES = {
  FREE: 0,
  KEEN: 1,
  BUSY: -1
};

export const AVAILABILITY_STATUSES_NAMES = {
  "0": "Free",
  "1": "Keen",
  "-1": "Busy"
};

export type AvailabilityStatus = -1 | 0 | 1;

class Availability {
  @observable status: AvailabilityStatus;

  constructor(availability?: AvailabilityStatus) {
    this.status = availability || AVAILABILITY_STATUSES.FREE;
  }

  @computed
  get statusName(): AvailabilityStatus {
    return AVAILABILITY_STATUSES_NAMES[this.status];
  }

  @action
  toggle() {
    switch (this.status) {
      case AVAILABILITY_STATUSES.BUSY:
        this.status = AVAILABILITY_STATUSES.FREE;
        break;
      case AVAILABILITY_STATUSES.KEEN:
        this.status = AVAILABILITY_STATUSES.BUSY;
        break;
      default:
        this.status = AVAILABILITY_STATUSES.KEEN;
    }
  }

  @action
  setStatus(status: AvailabilityStatus) {
    this.status = status;
  }
}

export default Availability;
