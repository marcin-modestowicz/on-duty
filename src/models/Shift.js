// @flow
import { observable, action, computed } from "mobx";
import type User from "./User";

export const USERS_PER_SHIFT = 2;

export class Shift {
  @observable onDuty: User[];

  constructor(users: User[] = []) {
    this.onDuty = users;
  }

  @action
  addUser = (user: User) => {
    this.onDuty.push(user);
  };

  @action
  removeUser = (id: string) => {
    const index = this.onDuty.findIndex(user => user.id === id);
    this.onDuty.splice(index, 1);
  };

  toggleUser = (user: User) => {
    const hasUser = this.onDuty.some(({ id }) => id === user.id);

    if (hasUser) {
      this.removeUser(user.id);
    } else {
      this.addUser(user);
    }
  };

  fill = (users: User[]) => {
    const newUsersCount = USERS_PER_SHIFT - this.onDuty.length;
    const usersToAdd = users.slice(0, newUsersCount);

    for (let i = 0; i < usersToAdd.length; i += 1) {
      this.addUser(users[i]);
    }
  };

  @computed
  get isReady(): boolean {
    return this.onDuty.length === USERS_PER_SHIFT;
  }
}

export default Shift;
