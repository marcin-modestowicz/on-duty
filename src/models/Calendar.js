// @flow
import { observable, computed, action } from "mobx";
import type User from "./User";
import Availability, { AVAILABILITY_STATUSES } from "./Availability";
import Shift, { USERS_PER_SHIFT } from "./Shift";

export const MINIMUM_REST_DAYS = 1;

export function daysInCurrentMonth() {
  const currentDate = new Date();
  const year = currentDate.getFullYear();
  const month = currentDate.getMonth() + 1;

  return new Date(year, month, 0).getDate();
}

export class ShiftsCalendar {
  @observable days: Shift[] = [];
  @observable summary: { [key: string]: string | number };

  constructor() {
    const maxDay = daysInCurrentMonth();

    for (let day = 1; day <= maxDay; day += 1) {
      this.days.push(new Shift());
    }
  }

  getUserShifts(userId: string): number[] {
    return this.days.reduce((sum, { isSealed, onDuty }, index) => {
      if (isSealed && onDuty.some(user => user.id === userId)) {
        sum.push(index);
      }
      return sum;
    }, []);
  }

  fillCalendar(users: User[]) {
    // Fill all slots in shift with users
    this.days.forEach((shift, index) => {
      users.forEach(user => {
        shift.addUser(user);
      });
    });

    // Create index array to iterate over, sort by preference
    const indexes = this.days
      .slice()
      .map((shift, index) => {
        return {
          index,
          preference: shift.onDuty.reduce(
            (sum, user) =>
              sum +
              (user.availabilityCalendar.days[index].status ===
              AVAILABILITY_STATUSES.KEEN
                ? user.power
                : 0),
            0
          )
        };
      })
      .sort((shiftA, shiftB) => {
        return shiftB.preference - shiftA.preference;
      })
      .map(shift => shift.index);

    // Iterate over shifts, sort users in shift
    indexes.forEach(index => {
      const currentShift = this.days[index];
      const notAllLimitsReached = users.some(
        user => this.getUserShifts(user.id).length < user.requiredShifts
      );

      // Sort users in shift
      currentShift._onDuty = currentShift._onDuty.sort((userA, userB) => {
        const userASortValue = this.getUserSortValue(
          userA,
          index,
          notAllLimitsReached
        );
        const userBSortValue = this.getUserSortValue(
          userB,
          index,
          notAllLimitsReached
        );

        return userASortValue - userBSortValue;
      });

      // Seal shift
      currentShift.seal();
    });

    // Create summary
    this.summary = this.createSummary(users);
  }

  getUserSortValue(
    user: User,
    shiftIndex: number,
    notAllLimitsReached: boolean
  ): number {
    const userShifts = this.getUserShifts(user.id);
    const userShiftCount = userShifts.length;
    const hasUserMinimumRestDays = !(
      userShifts.some(
        index => shiftIndex >= index - MINIMUM_REST_DAYS && shiftIndex < index
      ) ||
      userShifts.some(
        index => shiftIndex <= index + MINIMUM_REST_DAYS && shiftIndex > index
      )
    );
    const hasUserRequiredShiftsFilled = userShiftCount === user.requiredShifts;
    const userAvailabilityStatus =
      user.availabilityCalendar.days[shiftIndex].status;

    let userSortValue = -userAvailabilityStatus * user.power;

    if (!hasUserMinimumRestDays) {
      userSortValue += 20;
    }

    if (hasUserRequiredShiftsFilled && notAllLimitsReached) {
      userSortValue += 1;
    }

    return userSortValue;
  }

  createSummary(users: User[]) {
    const summary = users.reduce((allUsers, currentUser, index) => {
      const availabilityDays = currentUser.availabilityCalendar.days;
      const shiftDays = this.getUserShifts(currentUser.id);
      const highPreferenceDays = availabilityDays.reduce((sum, day, index) => {
        if (day.status === AVAILABILITY_STATUSES.KEEN) {
          sum.push(index);
        }

        return sum;
      }, []);
      const highPreference = highPreferenceDays.length;
      const highPreferenceFilled = highPreferenceDays.filter(index =>
        shiftDays.includes(index)
      ).length;
      const lowPreferenceDays = availabilityDays.reduce((sum, day, index) => {
        if (day.status === AVAILABILITY_STATUSES.BUSY) {
          sum.push(index);
        }

        return sum;
      }, []);
      const lowPreference = lowPreferenceDays.length;
      const lowPreferenceFilled = lowPreferenceDays.filter(index =>
        shiftDays.includes(index)
      ).length;
      const satisfaction =
        (highPreferenceFilled === 0
          ? 0
          : highPreferenceFilled / highPreference) -
        (lowPreferenceFilled === 0 ? 0 : lowPreferenceFilled / lowPreference);

      allUsers[currentUser.id] = {
        id: currentUser.id,
        name: currentUser.name,
        power: currentUser.power,
        highPreference,
        highPreferenceFilled,
        lowPreference,
        lowPreferenceFilled,
        shifts: shiftDays.length,
        shiftsShare: `${shiftDays.length /
          (this.days.length * USERS_PER_SHIFT / users.length) *
          100}%`,
        satisfaction
      };

      return allUsers;
    }, {});

    return summary;
  }
}

export class AvailabilityCalendar {
  days: Availability[] = [];

  constructor(availability?: number[] = []) {
    const maxDay = availability.length
      ? availability.length
      : daysInCurrentMonth();

    for (let day = 1; day <= maxDay; day += 1) {
      // $FlowFixMe
      this.days.push(new Availability(availability[day - 1]));
    }
  }
}
