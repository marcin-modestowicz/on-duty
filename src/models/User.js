// @flow
import { AvailabilityCalendar } from "./Calendar";
import { type AvailabilityStatus } from "./Availability";
import uuid from "uuid/v1";

export const POWER = {
  RESIDENT: 0,
  DOCTOR: 1,
  SPECIALIST: 2
};

const REQUIRED_SHIFTS = {
  RESIDENT: 2,
  DOCTOR: 1,
  SPECIALIST: 1
};

class User {
  id: string;
  name: string;
  isDoctor: boolean;
  isSpecialist: boolean;
  isAdmin: boolean;
  availabilityCalendar: AvailabilityCalendar;

  constructor(
    name: string,
    isDoctor?: boolean,
    isSpecialist?: boolean,
    availability?: AvailabilityStatus[]
  ) {
    this.id = uuid();
    this.name = name;
    this.isDoctor = !!isDoctor;
    this.isSpecialist = !!isSpecialist;
    this.isAdmin = false;
    this.availabilityCalendar = new AvailabilityCalendar(availability);
  }

  get power(): number {
    let power = POWER.RESIDENT;

    if (this.isDoctor) {
      power += POWER.DOCTOR;
    }

    if (this.isSpecialist) {
      power += POWER.SPECIALIST;
    }

    return power;
  }

  get requiredShifts(): number {
    let requiredShifts = REQUIRED_SHIFTS.RESIDENT;

    if (this.isDoctor) {
      requiredShifts = REQUIRED_SHIFTS.DOCTOR;
    }

    if (this.isSpecialist) {
      requiredShifts = REQUIRED_SHIFTS.SPECIALIST;
    }

    return requiredShifts;
  }
}

export default User;
