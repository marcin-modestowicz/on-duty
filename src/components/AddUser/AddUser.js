// @flow
import React, { Component } from "react";
import { observable, action } from "mobx";
import { observer } from "mobx-react";
import { type UserType, USER_TYPES } from "../../models/User";

/* global SyntheticInputEvent */

type Props = {
  onAdd: (name: string, type: UserType) => void
};

@observer
class AddUser extends Component<Props> {
  @observable userName: string = "";
  @observable userType: UserType = USER_TYPES.RESIDENT;

  @action
  handleNameChange = (event: SyntheticInputEvent<*>) => {
    this.userName = event.target.value;
  };

  @action
  handleTypeChange = (event: SyntheticInputEvent<*>) => {
    //$FlowFixMe
    this.userType = event.target.value;
  };

  @action
  handleUserAdd = () => {
    if (this.userName !== "") {
      this.props.onAdd(this.userName, this.userType);
      this.userName = "";
      this.userType = USER_TYPES.RESIDENT;
    }
  };

  render() {
    const isAddButtonDisabled = this.userName.trim() === "";
    return (
      <div>
        <input
          type="text"
          value={this.userName}
          onChange={this.handleNameChange}
        />
        <select value={this.userType} onChange={this.handleTypeChange}>
          {Object.keys(USER_TYPES).map(userType => (
            <option key={userType} value={USER_TYPES[userType]}>
              {USER_TYPES[userType]}
            </option>
          ))}
        </select>
        <button onClick={this.handleUserAdd} disabled={isAddButtonDisabled}>
          Add User
        </button>
      </div>
    );
  }
}

export default AddUser;
