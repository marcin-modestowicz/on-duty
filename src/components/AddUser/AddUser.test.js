// @flow
import React from "react";
import { shallow } from "enzyme";
import toJson from "enzyme-to-json";
import { USER_TYPES } from "../../models/User";
import AddUser from "./AddUser";

describe("AddUser component", () => {
  let onAdd;
  let addUser;
  let instance;

  beforeEach(() => {
    onAdd = jest.fn();
    addUser = shallow(<AddUser onAdd={onAdd} />);
    instance = addUser.instance();
  });

  test("should render with button disabled", () => {
    expect(toJson(addUser)).toMatchSnapshot();
  });

  test("should render with button enabled", () => {
    instance.handleNameChange({ target: { value: "Marty McFly" } });
    expect(toJson(addUser)).toMatchSnapshot();
  });

  describe("handleNameChange method", () => {
    test("should update user name", () => {
      instance.handleNameChange({ target: { value: "Batman" } });
      expect(instance.userName).toBe("Batman");
    });
  });

  describe("handleTypeChange method", () => {
    test("should update user type", () => {
      instance.handleTypeChange({ target: { value: USER_TYPES.DOCTOR } });
      expect(instance.userType).toBe(USER_TYPES.DOCTOR);
    });
  });

  describe("handleUserAdd method", () => {
    describe("if user name is not filled", () => {
      test("should do nothing", () => {
        instance.handleUserAdd();
        expect(onAdd).not.toHaveBeenCalled();
      });
    });

    describe("if user name is filled", () => {
      beforeEach(() => {
        instance.handleNameChange({ target: { value: "Bronco Billy" } });
        instance.handleUserAdd();
      });

      test("should call onAdd prop", () => {
        expect(onAdd).toHaveBeenCalledWith("Bronco Billy", USER_TYPES.RESIDENT);
      });

      test("should reset user name and type", () => {
        expect(instance.userName).toBe("");
        expect(instance.userType).toBe(USER_TYPES.RESIDENT);
      });
    });
  });
});
