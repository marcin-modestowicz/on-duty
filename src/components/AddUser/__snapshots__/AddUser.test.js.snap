// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddUser component should render with button disabled 1`] = `
<div>
  <input
    onChange={[Function]}
    type="text"
    value=""
  />
  <select
    onChange={[Function]}
    value="Resident"
  >
    <option
      key="RESIDENT"
      value="Resident"
    >
      Resident
    </option>
    <option
      key="DOCTOR"
      value="Doctor"
    >
      Doctor
    </option>
    <option
      key="SPECIALIST"
      value="Specialist"
    >
      Specialist
    </option>
  </select>
  <button
    disabled={true}
    onClick={[Function]}
  >
    Add User
  </button>
</div>
`;

exports[`AddUser component should render with button enabled 1`] = `
<div>
  <input
    onChange={[Function]}
    type="text"
    value="Marty McFly"
  />
  <select
    onChange={[Function]}
    value="Resident"
  >
    <option
      key="RESIDENT"
      value="Resident"
    >
      Resident
    </option>
    <option
      key="DOCTOR"
      value="Doctor"
    >
      Doctor
    </option>
    <option
      key="SPECIALIST"
      value="Specialist"
    >
      Specialist
    </option>
  </select>
  <button
    disabled={false}
    onClick={[Function]}
  >
    Add User
  </button>
</div>
`;
